<!doctype html>
<html>

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>EpiAlignment Manual</title>
  <link href='https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Material+Icons' rel="stylesheet" type="text/css">
  <link href="styles/mainstyles.css" rel="stylesheet" type="text/css">
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-78231712-3"></script>
  <script src="js/googleAnalytics.js"></script>
  <script defer src="https://cdn.jsdelivr.net/npm/vue@2.5.21/dist/vue.min.js"></script>
  <script defer src="js/manual.js" type="module"></script>
</head>

<body class="manual">
  <div id="manual_app" class="flex flexRow flexNoGap flexNoWrap" v-cloak>
    <div v-show="!showNav" id="callOutNav" @click="toggleNav(true)">
      Table of contents
    </div>
    <transition name="nav-slide-right">
      <div v-show="showNav" id="manualNav" class="noFlex flexColumn flexNoGap flexEnd">
        <div id="navHeader" class="flexRow flexWrap flexNoGap flexCenter">
          <div class="flex">Table of contents</div>
          <div id="collapseNavButton" @click="toggleNav(false)">Collapse</div>
        </div>
        <div id="navContent" class="flex">
          <ul>
            <li>
              <a href="#introduction" @click="clickOnNav">Introduction</a>
            </li>
            <li>
              <a href="#quickStart" @click="clickOnNav">Quick start</a>
              <ul>
                <li>
                  <a href="#oneVsOneExample" @click="clickOnNav">One-vs-one mode example</a>
                </li>
                <li>
                  <a href="#manyVsManyExample" @click="clickOnNav">Many-vs-many mode example</a>
                </li>
              </ul>
            </li>
            <li>
              <a href="#usage" @click="clickOnNav">Usage</a>
              <ul>
                <li>
                  <a href="#usageOverview" @click="clickOnNav">Overview</a>
                </li>
                <li>
                  <a href="#oneVsOneMode" @click="clickOnNav">One-vs-one mode</a>
                </li>
                <li>
                  <a href="#manyVsManyMode" @click="clickOnNav">Many-vs-many mode</a>
                </li>
              </ul>
            </li>
            <li>
              <a href="#input" @click="clickOnNav">Input</a>
              <ul>
                <li>
                  <a href="#genomeAssembly" @click="clickOnNav">Genome assembly</a>
                </li>
                <li>
                  <a href="#peakFileAndDatasets" @click="clickOnNav">Peak files and public datasets</a>
                </li>
                <li>
                  <a href="#queryAndTargetRegions" @click="clickOnNav">Query and target regions</a>
                  <ul>
                    <li>
                      <a href="#oneVsOneModeQueryAndTarget" @click="clickOnNav">One-vs-one mode</a>
                    </li>
                    <li>
                      <a href="#manyVsManyModeQueryAndTarget" @click="clickOnNav">Many-vs-many mode</a>
                    </li>
                  </ul>
                </li>
                <li>
                  <a href="#parameters" @click="clickOnNav">Parameters</a>
                  <ul>
                    <li>
                      <a href="#weight" @click="clickOnNav">The epigenome weight</a>
                    </li>
                    <li>
                      <a href="#alignmentParameters" @click="clickOnNav">Alignment parameters</a>
                    </li>
                  </ul>
                </li>
              </ul>
            </li>
            <li>
              <a href="#output" @click="clickOnNav">Output</a>
              <ul>
                <li>
                  <a href="#oneVsOneOutput" @click="clickOnNav">One-vs-one mode</a>
                </li>
                <li>
                  <a href="#manyVsManyOutput" @click="clickOnNav">Many-vs-many mode</a>
                </li>
              </ul>
            </li>
            <li>
              <a href="#pythonClientTemplate" @click="clickOnNav">Python client template</a>
            </li>
            <li>
              <a href="#furtherReading" @click="clickOnNav">Further reading</a>
            </li>
            <li>
              <a href="#references" @click="clickOnNav">References</a>
            </li>
          </ul>
        </div>
      </div>
    </transition>
    <div id="manualContent" ref="manualContent" class="flex flexColumn flexNoWrap">
      <div id="header" class="noFlex flexRow flexCenter flexWrap">
        <!-- Introduction -->
        <div id="headerContainer" class="flex flexColumn flexStretch flexNoWrap flexNoGap">
          <h1 class="flex"><a href="index.html">EpiAlignment: a chromosomal similarity search engine</a></h1>
          <div class="headerByline">
            EpiAlignment is a tool that can align genomic regions with both sequence
            and epigenomic information.
          </div>
        </div>
        <div id="headerButtonHolder" class="flex flexRow flexWrap flexCenter justifyEnd">
          <a class="buttons flexRow flexCenter justifyCenter" title="A template python program that can be used to access EpiAlignment"
            target="_blank" href="assets/PythonClient.zip"><i class="material-icons">save_alt</i>Python
            client template</a>
        </div>
      </div>
      <div ref="introduction" class="manualHeader">
        <h1>EpiAlignment documentation</h1>
      </div>
      <div class="section">
        <!--Introduction-->
        <h1>Introduction</h1>
        <p>EpiAlignment is a web-based tool that can align genomic regions with both sequence and epigenomic information.
           This web server aims to assist users with matching genomic elements
           with functional and structural similarities across two species.</p>
        <p>The web server has two alignment modes: one-vs-one (default) and many-vs-many. The default one-vs-one mode is 
           suitable if you have <strong>medium-sized query regions, 
           usually marked by epigenomic modifications, (e.g. ChIP-Seq peaks, hundreds to thousands of base pairs long)</strong>, 
           and much longer target regions. 
           It will match the query region(s) to a sub-region of the target region(s). The many-vs-many mode attempts to align 
           every input query region to all target regions to find the best hit. 
           It would be useful if you have query and target regions of similar lengths (medium-sized), and would like to find one best match 
           for the query region among a set of target regions. The two modes are illustrated in figures below.</p>
        <div class="flexRow">
          <div class="manualImg small">
            <img alt="Expandable image for one-vs-one mode" src="assets/images/DefaultStart.png">
          </div>
          <div class="manualImg small">
            <img alt="Expandable image for many-vs-many mode" src="assets/images/ManyStart.png">
          </div>
        </div>
      </div>
      <div ref="quickStart" class="section">
        <!--Quick Start-->
        <h1>Quick start</h1>
        <p>EpiAlignment has two major alignment modes: one-vs-one (default) and many-vs-many. Given an input query region 
          (usually hundreds to thousands of bp), the default
          mode can be used to identify the best-matching sub-region within the corresponding target region, 
          whereas the many-vs-many mode is useful for finding the best match among a set of target regions.</p>

        <h3 ref="oneVsOneExample">An example for using the one-vs-one mode</h3>
        <!--Example 1-->
        <p>In this example, we will search a H3K4me3-marked mouse genomic region (~2kb) against the neighborhood of its
          homologous region in human. This example will take about one minute to run on EpiAlignment.</p>
        <h4>Import sample data with one click:</h4>
        <div class="exampleSteps">
          <ul>
            <li>Make sure that the "One-vs-one" mode is selected in STEP 1.</li>
            <li>Click the "Use sample data" button in STEP 2.</li>
            <li>Click "Submit" in STEP 4</li>
          </ul>
        </div>

        <h4>OR, enter data manually:</h4>
        <div class="exampleSteps">
          <ul class="sampleSteps">
            <li>Make sure that the "One-vs-one" mode is selected in STEP 1.</li>
            <li>In STEP 2, "Query region" box:</li>
            <ul class="sampleStepsSub">
              <li>Set "Genome assembly" to mm10.</li>
              <li>Press "Click here to select a paired public dataset", find "Public Datasets" and select "Round
                spermatids
                ChIP-Seq (H3K4me3)" by clicking on the blue cell.
                Press the confirm button. </li>
              <li>Paste the following mouse genomic region into the "Input regions" textarea:</li>
              <div>
                <pre>chr12&#9;8207583&#9;8209349&#9;Region_example&#9;0&#9;+</pre>
              </div>
            </ul>
            <li>In the "Target region" box, set "Genome assembly" to hg38.</li>
            <li>Make sure that "Use homologous regions in this species" is selected. This is the default mode we are
              going
              to use.</li>
            <li>Leave the size of bps upstream as 20000 and change the size of bps downstream to 0.</li>
            <li>Click "Submit" in STEP 4.</li>
          </ul>
        </div>

        <h4>Output</h4>
        <p>You will be redirected to the result page after submitting the data. In this mode, the result page
          shows
          an interactive table with details of the alignments. You can click on each row to expand it and view the alignment
          scores yielded with and without epigenomic information along the target region. For visualization, 
          you can click the icon (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">GIVE</span>
          </span>)
          at the beginning of each row to view query and target regions together 
          in the Genomic Interaction Visualization Engine (GIVE), or view each region individially in the UCSC genome 
          browser by clicking the icon (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">UCSC</span>
          </span>)
          beside each coordinate. In the UCSC genome browser, you can
          select any preset ENCODE/public epigenomic data from the custom tracks.
          </p>
        <div class="manualImg large">
          <img alt="Expandable data table for one-vs-one mode" src="assets/images/Enhancer_result_comment.png">
        </div>
        <p>The table can be downloaded by clicking "Download results".</p>

        <h3 ref="manyVsManyExample">An example for using the many-vs-many mode</h3>
        <!--Example 2-->
        <p>In this example, we will use gene promoters as input regions. We will search the promoter of 
          a human protocadherin beta gene, PCDHB5,
          against a mouse gene cluster to find mouse promoters with the
          highest structural similarities
          within it. The mouse gene cluster contains promoters of six mouse protocadherin beta genes: Pcdhb4, Pcdhb6,
          Pcdhb8, Pcdhb10, Pcdhb11 and Pcdhb12, all of which are annotated orthologous genes of PCDHB5.</p>
        <h4>Import sample data with one click:</h4>
        <div class="exampleSteps">
          <ul>
            <li>Select the "Many-vs-many" alignment mode in STEP 1.</li>
            <li>Click the "Use sample data" button in STEP 2.</li>
            <li>Click "Submit" in STEP 4</li>
          </ul>
        </div>
        <h4>OR, enter data manually:</h4>
        <div class="exampleSteps">
          <ul class="sampleSteps">
            <li>Select the "Many-vs-many" alignment mode in STEP 1.</li>
            <li>In STEP 2, "Query region" box:</li>
            <ul class="sampleStepsSub">
              <li>Press "Click here to select a paired public dataset", find "Public Datasets" and select "Round
                spermatids
                ChIP-Seq (H3K4me3)" by clicking on the blue cell.
                Press the confirm button. </li>
              <li>Paste or type the following gene name into the "Input regions"
                textarea, <em>one name per line</em>:</li>
              <div>
                PCDHB5
              </div>
            </ul>
            <li>In the "Target region" box, make sure that "Search a gene cluster"
              is selected, and type "PCDHB5" in the search bar. In the pop-up panel,
              click the box with arrow
              (<i class="material-icons">open_in_browser</i>) to the right of
              the first listed gene cluster <strong>"Cluster_1491"</strong> to select it.</li>
            <li>Click "Submit" in STEP 4.</li>
          </ul>
        </div>

        <h4>Output</h4>
        <p>After successfully submitting the data, you will be redirect to the result page. In this mode, 
          two heatmaps and a table will be retunred. The heatmaps illustrate alignment scores of every promoter pair 
          and the best hits found for each promoter. The three digits behind each gene name is the id
          of the promoter region. </p>
        <div class="manualImg large">
          <img alt="Heatmap for many-vs-many mode" src="assets/images/Heatmap.png">
        </div>
        <p>The table shows details of the alignments between each pair of promoters. You can sort the table by each column
          by clicking the column name. Click the arrow beside "#" to reset the order. 
          Similar to that in the default mode, You can click the <i class="material-icons">open_in_new</i> 
          icons to view the regions
          in the UCSC genome browser or GIVE. The table can be
          downloaded by clicking "Download results".</p>
        <div class="manualImg large">
          <img alt="Data table for many-vs-many mode" src="assets/images/PromoterTable_comments.png">
        </div>
        <p>EpiAlignment also provides many other ways of defining query and target regions. Please check out the Usage
          section for more information.</p>
      </div>
      <div ref="usage" class="section">
        <!--Usage-->
        <h1 ref="usageOverview">Usage</h1>
        <h2>Overview</h2>
        <p>EpiAlignment provides two major alignment modes: one-vs-one and many-vs-many. Both modes use a semi-global
          approach to
          identify the best alignment between a pair of regions, whereas the result presentation and interpretation are
          different. 
          The one-vs-one mode attempts to find a best-matching part for a query region in the 
          corresponding target region (one-against-one alignment). 
          It is more useful in cases where query regions are functional genomic elements marked by epigenomic modifications (usually hundreds 
          to thousands base pairs), while the target regions 
          are much longer,
          and the best match of the query region cannot be easily determined in the target species with sequence homologies only.
          This mode allows defining target regions using homologous information.</p>
        <p>The many-vs-many mode, on the other hand, aligns every input query region to 
          all target regions to find the best
          hit (all-against-all alignment). Therefore, it would be more useful in cases where the query region corresponds to 
          multiple candidates with potentially similar structures and functions, such as promoters of paralogous genes. This
          mode could also be handy if you'd like to use gene promoters as input regions 
          as you can define promoter regions with gene names. We recommend users to provide query and target
          regions with similar lengths in this mode.</p>
        <p>Each alignment mode provides various ways of defining target regions for users' convenience. 
          An overview of the alignment modes
          and
          target region types supported is shown below:</p>
        <div>
          <table>
            <caption>Alignment modes and supported target region types</caption>
            <thead>
              <tr>
                <th>Alignment mode</th>
                <th>Supported target region types</th>
                <th>Description</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <th rowspan="2">One-vs-one mode</th>
                <td>Homologous regions</td>
                <td>For each query region, search for the best-matching sub-region in the neighborhood of its homologous region.</td>
              </tr>
              <tr>
                <td>Genomic coordinates</td>
                <td>For each query region, search for the best-matching sub-region within a designated target region.</td>
              </tr>
              <tr>
                <th rowspan="2">Many-vs-many mode</th>
                <td>A gene cluster</td>
                <td>Align each query region against all isoform promoters in a group of 
                evolutionarily-related genes to identify the best-matching promoter.</td>
              </tr>
              <tr>
                <td>Genomic coordinates or gene names</td>
                <td>Align each query region against all target regions to identify the best match.</td>
              </tr>
            </tbody>
          </table>
        </div>

        <h3 ref="oneVsOneMode">One-vs-one mode (default)</h3>
        <h4>Use homologous regions in this species</h4>
        <p>This mode allows users to define target regions using homologous information. Specifically, EpiAlignment
          will first extend each query region by a user-specified distance (default: 20,000kb) upstream and downstream, and then
          use <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver" target="_blank">liftOver</a> to remap the extended
          region to the target species. If the remapping succeeds, the result region will be used as the target region.</p>
        <h4>Define target regions with a BED file</h4>
        <p>By select this option, users can provide their own
          query and target regions. EpiAlignment will pair the query targe regions in the two input files based on
          their
          line number and perform a one-against-one alignment to identify the best-matching sub-regions.</p>

        <h3 ref="manyVsManyMode">Many-vs-many mode</h3>
        <h4>Search a gene cluster</h4>
        <p>This mode allows users to align query regions to promoters of a group of evolutionarily-related genes and find
          the
          best match. The gene cluster data is preset in EpiAlignment, identified by grouping paralogous genes in each
          species and linking orthologous genes across species. Different isoform promoters of the same genes
          are
          all included. An example of a pre-identified gene cluster is shown below, in which <em>Aph1a</em>, <em>Aph1b</em>
          and <em>Aph1c</em> are paralogues in mouse, and <em>APH1A</em>, <em>APH1B</em> are paralogues in human.</p>
        <div>
          <table class="verticalBorder">
            <caption>An example of a pre-identified gene cluster</caption>
            <thead>
              <tr>
                <th></th>
                <th><em>Aph1a</em></th>
                <th><em>Aph1b</em></th>
                <th><em>Aph1c</em></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <th><em>APH1A</em></th>
                <td>Orthologous</td>
                <td></td>
                <td></td>
              </tr>
              <tr>
                <th><em>APH1B</em></th>
                <td></td>
                <td>Orthologous</td>
                <td>Orthologous</td>
              </tr>
            </tbody>
          </table>
        </div>
        <p>By selecting this cluster, you can either align any human genomic region(s) to the promoters of mouse genes
          <em>Aph1a</em>,
          <em>Aph1b</em>, <em>Aph1c</em>, or any mouse genomic region(s) to the promoters of human genes <em>APH1A</em>,
          <em>APH1B</em>.</p>

        <h4>Define target regions with a BED file / a gene list</h4>
        <p>Users can define their own query and target regions by selecting this option. Genomic coordinates, 
          gene symbols and Ensembl IDs
          are accepted as inputs. If gene names are provided, EpiAlignment will automatically fetch all
          promoters
          of the genes, which could be handy if you don't want to retrieve promoter regions by yourself. EpiAlignment
          will
          then align each input query region to all input target regions to find top hits.</p>

        <h2 ref="input">Input</h2>
        <!--Input-->
        <p>EpiAlignment requires two types of inputs from users: (1) a pair of peak files from ChIP-Seq experiments, 
          and (2) query and target regions to be aligned, both from two species. 
          The input file formats are different depending on the alignment mode selected.</p>
        <h3 ref="genomeAssembly">Genome assembly</h3>
        <p>Genome assemblies of the query and target regions should first be specified using the dropdown shown below:</p>
        <div class="manualImg small">
          <img alt="Dropdown for genome assemblies" src="assets/images/genomeAssembly.png">
        </div>
        <p>EpiAlignment currently supports two genome assemblies: hg38 (Homo sapiens) and mm10 (Mus musculus). Please
          note
          that the genome assemblies for the query and target regions have to be different if you choose to define
          target
          regions using homologous regions in the default mode. For other modes, the two genome assemblies can be
          identical whereas not recommended.</p>

        <h3 ref="peakFileAndDatasets">Peak files and datasets for pairwise ChIP-Seq experiments</h3>
        <p>EpiAlignment provides users with comparable epigenomic datasets of matching tissue/cell types, curated from the RoadMap Epigenomics 
           and ENCODE consortia and public comparative epigenomic studies.To select peak files, 
           in the Query regions box, click "Click here to select a paired public dataset."</p>
        <div class="manualImg">
          <img alt="Public dataset selection button" src="assets/images/presetDataButton.png">
        </div>
        <p>In the pop-up window, click the cell representing the tissue type and histone mark that you want to select.
          There are two numbers on each cell in the format of "H:x M:y". The two numbers represents number of peak
          files
          available in human and mouse, respectively. Once you click the cell, available experiment IDs will be listed
          on
          the
          rightside. You may click the icon beside the experiment ID to see experiment details on their original
          ENCODE/GEO
          database page.</p>
        <p>Click on the experiment ID you'd like to select, then press confirm. The selected experiment
          will display in the Query regions and Target regions boxes afterwards.</p>
        <p>You may also click anywhere outside the pop-up window at any time to close it without making any changes to
          your
          peak file selection.
        </p>
        <div class="manualImg">
          <img alt="Public dataset selection window" src="assets/images/presetWindow.png">
        </div>
        <p>The user can also upload their own peak files. To do this, choose and upload your peak file 
          (<a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1" target="_blank">BED6</a> format)
          for
          the first species in the Query regions box, and the peak file for the second species in the Target regions
          box.
          <em>Uploading your own peak file will clear any previously selected public data in all references.</em></p>
        <div class="manualImg small">
          <img alt="Upload your own peak file" src="assets/images/UploadPeak_comment.png">
        </div>

        <h3 ref="queryAndTargetRegions">
          Query and target regions
        </h3>
        <h4 ref="oneVsOneModeQueryAndTarget">One-vs-one mode</h4>
        <h5>Query regions</h5>
        <p>For the default mode, the input query regions need to be defined with genomic coordinates in BED6 format.
          The coordinates can be either pasted into the textarea (one region per line) or uploaded as a file. Typically, the 
          query regions are hundreds to thousands bp-long genomic elements marked by epigenomic modifications, such as ChIP-Seq 
          peaks. If your query regions are longer than 10kb, a warning will show up on the result page.
          Please note that each line should
          have a unique name. A sample data is shown below:</p>
        <div class="manualImg">
          <img alt="One-vs-one mode sample query regions" src="assets/images/Enhancer_s1.png">
        </div>
        <h5 ref="oneVsOneModeTarget">Target regions</h5>
        <h6>Use homologous regions in this species</h6>
        <p>In this case, users only need to specify the number of base pairs 
          by which the query regions will be increased. By default, they will be extended 20,000 kb
          from the two ends respectively.</p>
        <div class="manualImg small">
          <img alt="Use homologous regions for enhancer targets" src="assets/images/OrthoMode_comments.png">
        </div>

        <h6>Define target regions with a BED file</h6>
        <p>In this mode, target should be provided as genomic coordinates in BED6 format. The list should have the
          same row count and order as the query regions, as EpiAlignment will directly pair them based on their
          line
          number, and align each pair separately. EpiAlignment will only use names of the query regions to
          refer to individual pairs, so it is recommended to keep the names same between query and the target
          regions. A sample input file is shown below, paired with the sample query region file shown above:</p>
        <div class="manualImg">
          <img alt="One-vs-one mode sample target regions" src="assets/images/Enhancer_s2.png">
        </div>


        <h4 ref="manyVsManyModeQueryAndTarget">Many-vs-many mode</h4>
        <h5>Query regions</h5>
        <p>In this mode, the query regions can be provided as genomic coordinates in BED6 format. 
          Gene symbols or Ensembl IDs are also accepted if gene promoters are used as input regions. 
          The coordinates and gene names can be either
          pasted into the textarea (one region/name per line) or uploaded as a file. 
          When using gene symbols or Ensembl IDs, ranges of promoters need to be specified. 
          By default, the (-1,000, +500) flanking regions around TSSs are used as promoter regions. Please note that 
          gene symbols are case-sensitive. For human gene symbols, all letters should be uppercase,
          whereas for mouse gene symbols, only the first letter should be uppercase.
          Sample query regions are shown below: </p>
        <div class="manualImg">
          <img alt="Sample files for promoter query regions" src="assets/images/sampleGeneName.png">
        </div>
        <div class="manualImg">
          <img alt="Select TSS flanking ranges for promoters" src="assets/images/TSSrange_comments.png">
        </div>
        <p>Given that a gene may have transcripts with overlapping promoter regions, EpiAlignment will merge promoters
          of the same gene if their overlapping region is longer than 80% of the promoter length. For example, if
          promoter regions are defined as 500kb upstream and 1000kb downstream the TSS, two promoters will be merged if
          their overlapping region is longer than <span class="equation">1500 &#215; 0.8 = 1200</span>. The last three
          digits of the promoter's genomic coordinate will be used as promoter id.</p>
        <h5 ref="manyVsManyModeTarget">Target regions</h5>
        <h6>Search a gene cluster</h6>
        <p>In this mode, you may search for a gene cluster and use promoters of genes in the cluster as target regions.
          You
          can type an Ensembl ID or a gene symbol in the searching bar. A search results panel will pop up with
          candidate
          clusters listed in it. Please note that gene without matching
          paralogues
          in the species are not included in any cluster. 
          You may click on the cluster name to further examine genes and their descriptions in the
          cluster. Click the box with arrow (<i class="material-icons">open_in_browser</i>) on the rightside of the
          cluster name to select the cluster you would like to
          use. You may click anywhere outside the pop-up panel at anytime to close it.</p>
        <div class="manualImg large">
          <img alt="Gene cluster search and selection panel" src="assets/images/GeneClusters_comment.png">
        </div>

        <h6>Define target regions with a BED file / a gene list.</h6>
        <p>In this case, the
          format of the input target regions is exactly the same as query regions. The query and target region files do
          not need to have identical row counts, as EpiAlignment will perform an all-against-all alignment.</p>


        <h3 ref="parameters">Parameters</h3>
        <p>The alignment algorithm of EpiAlignment is developed on an evolutionary basis [<a href="#ref_1" @click="clickOnNav">1</a>].
          Therefore, the parameters describe evolutionary events. The user may change the parameters to adjust the
          rewards
          and penalties assigned to sequences and epigenomic modifications.</p>
        <h4 ref="weight">The epigenome weight</h4>
        <p> This parameter represents the relative weight of the epigenome to the sequence. The weight will be
          multiplied to the epigenome parts of the alignment scores. With a
          larger
          epigenomic weight, the alignment will rely more on epigenomic similarity than sequence
          similarity. Consequently, regions with larger sequence differences (i.e., more mismatches and indels) but
          similar
          epigenomic states may beat regions with similar sequences but different epigenomic states. Additionally, in
          this
          case, a region pair with a larger overlap between epigenomic peaks is more likely to gain a higher
          alignment score. If you would like to increase the default epigenomic weight, please carefully check the
          sequence
          similarities and the epigenomic occupancies in the best hits.</p>

        <h4 ref="alignmentParameters">Alignment parameters</h4>
        <ul id="paralist">
          <li><strong><em>s</em></strong>: sequence substitution rate. With a smaller <em>s</em>, a larger reward score
            will be assigned to a pair of matched bases, and a larger penalty score will be assigned to a pair of
            mismatched bases.</li>
          <li><strong><em>&#956;</em></strong>: sequence deletion rate. A larger gap penalty will be assigned to
            insertions
            and deletions with a smaller <em>&#956;</em>.</li>
          <li><strong><em>&#954;</em></strong>: rate of switching between the two epigenomic states 0 and 1. With a
            smaller
            <em>&#954;</em>, a larger reward score will be assigned to a pair of matched epigenomic states, and a
            larger
            penalty score will be assigned to a pair of mismatched epigenomic states.</li>
          <li><strong><em>&#960;<sub>A</sub>, &#960;<sub>C</sub>, &#960;<sub>G</sub> and &#960;<sub>T</sub>:</em></strong>equilibrium
            probabilities of the four nucleotide bases.</li>
          <li><strong><em>&#960;<sub>1</sub></em></strong>: equilibrium probability of the epigenomic state 1, that is,
            the
            probability of having the epigenomic modification on genomic locations.</li>
        </ul>

        <h2 ref="output">Output</h2>
        <p>You will be redirect to the result page after successfully submitting your data. The result page will
          refresh
          automatically until the task is done. To align a pair of regions of lengths <span class="equation"><em>m</em></span>
          and
          <span class="equation"><em>n</em></span>, <span class="equation">O(<em>mn</em>)</span> time is required by
          EpiAlignment. Therefore, given your region lengths and numbers, the task may take minutes to hours to finish.
          You
          may provide your email and choose to be notified by email after the task is done, or you can bookmark the
          page
          and come back later. We recommend that the users select either way if there are more than 100 region pairs,
          especially when using the one-vs-one mode.</p>

        <h3 ref="oneVsOneOutput">One-vs-one mode</h3>
        <p>For the one-vs-one mode, the result page shows an interactive table with alignment details. The table can be
          sorted by each column by clicking the arrow beside column name. Click the arrow beside
          "#" to reset the order. The table can be downloaded by clicking "Download results".</p>
        <p>You may click the icon (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
          <span class="annotation">GIVE</span>
        </span>) at the front of each row to launch a browser powered by GIVE to have a side-by-side view of both the query
          and the target regions. If you have chosen a preset dataset, the dataset will be included in the browser
          automatically. If you click the icon (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
          <span class="annotation">UCSC</span>
          </span>) beside each genome coordinates, you can
          see the location in UCSC genome browser. The bigWig files of all preset
          ENCODE and public datasets were available in this browser page. You may select any ENCODE/public epigenomic data
          tracks in "Custom Tracks", change the display mode from "hide" to "dense" to view them at the same location. </p>
        <p>Besides, each row of the table can be clicked to expand and show the details of the one-vs-one alignment. This
          includes highest alignment scores, and how the alignment scores
          with ("EpiAlign") or without ("SeqOnly") epigenomic information, distribute along the target region plotted in a
          graph.
          Note: if you set epigenomic weight to zero during input, the EpiAlignment score will be the degenerated
          sequence-only alignment score, and no "SeqOnly" result will be reported.</p>
        <div class="manualImg large">
          <img alt="Expandable data table for one-vs-one mode" src="assets/images/Enhancer_result_comment.png">
        </div>

        <h3 ref="manyVsManyOutput">Many-vs-many mode</h3>
        <p>For the many-vs-many mode, the result page shows two heatmaps and a table. EpiAlignment scores are plotted in
          the
          first heatmap to show scores of any query region (row) against all target region(s) (column). A sequence-only
          alignment result is also provided for comparative purposes, with scores illustrated in the second heatmap.
          The row and column names are region names. If gene names or gene clusters are used as the query input to align
          promoter regions, the row and column names will consist of gene names and the last three digits of
          promoters' genomic coordinates. In each row, the cell with the highest score of the row is marked with a
          capital "H".</p>
        <p>For each query region (row), EpiAlignment detects if its best hit switches from one target region to another
          when aligning with and without epigenomic information. The best hit is considered as "changed" if it meets
          the following criteria:</p>

        <div class="exampleSteps">
          <div class="flexRow flexWrap flexStretch">
            <table>
              <caption>Alignment scores with epigenomic information</caption>
              <thead>
                <tr>
                  <th></th>
                  <th>Target region A</th>
                  <th>Target region B</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th>Query region</th>
                  <td><span class="equation"><em>E<sub>A</sub></em></span></td>
                  <td><span class="equation"><em>E<sub>B</sub></em></span></td>
                </tr>
              </tbody>
            </table>
            <table>
              <caption>Alignment scores without epigenomic information</caption>
              <thead>
                <tr>
                  <th></th>
                  <th>Target region A</th>
                  <th>Target region B</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th>Query region</th>
                  <td><span class="equation"><em>S<sub>A</sub></em></span></td>
                  <td><span class="equation"><em>S<sub>B</sub></em></span></td>
                </tr>
              </tbody>
            </table>
          </div>
          <p>Assume that in alignment <em>without</em> epigenomic information,
            <span class="equation"><em>S<sub>A</sub></em> &#62; <em>S<sub>B</sub></em></span>,
            whereas in alignment <em>with</em> epigenomic information,
            <span class="equation"><em>E<sub>A</sub></em> &#60; <em>E<sub>B</sub></em></span>. For
            the query region, the best hit of the query
            region will be considered as switched from
            Target A to B when using EpiAlignment if <span class="equation"><em>E<sub>B</sub></em> &#62; <em>S<sub>B</sub></em></span>
            and <span class="equation"><em>E<sub>A</sub></em> &#60; <em>S<sub>A</sub></em></span>.
            In other words, we required that for the hit found by EpiAlignment with epigenomic information,
            the epigenomic information has to reinforce the chromosomal similarity, where as for the hit found by
            sequence-only alignment, the epigenomic information has to undermine the chromosomal similarity.</p>
        </div>

        <p>For query regions with switched target, the best hit found by EpiAlignment with epigenomic information
          is highlighted by a green box, and the best hit found by alignment without epigenomic information is
          highlighted by a gray box. </p>

        <p>The table shows details of the alignments between each pair of promoters. Similar to the one-vs-one mode,
          You may click the icon
          (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">GIVE</span>
          </span>) at the front of each row to see a side-by-side view in the GIVE browser, or click
          (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">UCSC</span>
          </span>)
          beside each genome coordinates to view the location in UCSC genome browser. </p>
        <p>You can also sort the table by each column by clicking the arrow beside column name, or download the table
          by clicking "Download results".</p>
        <div class="manualImg large">
          <img alt="Heatmap for many-vs-many mode" src="assets/images/Heatmap.png">
        </div>
      </div>

      <div ref="pythonClientTemplate" class="section">
        <h1>Python client template</h1>
        <p>We provided EpiAlignment_PyClient.py, a python template program for users to access EpiAlignment from their
          own
          server, submit multiple jobs and parse the output. The user may also modify the code and incorporate it into
          their own programs. The basic usage of the template program is:</p>
        <code>python EpiAlignment_PyClient.py sampleSheet.txt</code>
        <p>where sampleSheet.txt is a tab-delimited text file with details of the jobs you would like to submit to
          EpiAlignment.</p>
        <p>The python client template can be downloaded from the homepage of EpiAlignment. Sample input files are also
          available in the same zip file. Please follow <a href="https://github.com/Zhong-Lab-UCSD/EpiAlignment_WebApp/blob/master/html/assets/PythonClient/README.md"
            target="_blank">the README.md file</a> (also included in the zip file) to prepare your sample sheet and
          submit jobs to EpiAlignment.</p>
      </div>
      <!--<div class="section">
        <h1>Computational methods</h1>
        <p>EpiAlignment uses a dynamic programming algorithm for the comparison of sequences and epigenomic modifications. The input data is a pair of genomic regions: the query region <i>A</i> and the target region <i>B</i>, each of which consists of the DNA sequence and the epigenomic modifications on it. We assume that the state for each epigenomic modification on each genomic location is binary as shown in the figure below. On each base, a binary indicator (1,0) is used to represent the epigenomic state of an epigenomic mark, with the epigenomic state equal to 1 within regions marked by the epigenomic mark, and 0 outside such regions. </p>
    
        <p>EpiAlignment starts from scoring a pair of nucleoside bases from two sequences, rewarding for matches and penalizing for mismatches, gap openings or gap extensions. On top of each pair of bases, EpiAlignment adds a score from the epigenomic states to the alignment score, rewarding for matches and penalizing for mismatches of each epigenomic mark. Through dynamic programming, EpiAlignment computes alignment scores of any pairs of sub-sequences, and reports the path with the highest alignment score. </p>
        
      </div> -->

      <div ref="furtherReading" class="section">
        <!-- Further reading -->
        <h1>Further Reading</h1>
        <ol>
          <li>
            Xiao, S., Xie, D., Cao, X., Yu, P., Xing, X., Chen, C. C., ... &amp; Wang, T. (2012). Comparative epigenomic annotation of regulatory DNA. <em>Cell</em>, <strong>149</strong>(6), 1381-1392.
          </li>
          <li>
            Xiao, S., Cao, X., &amp; Zhong, S. (2014). Comparative epigenomics: defining and utilizing epigenomic variations across species, time‐course, and individuals. <em>Wiley Interdisciplinary Reviews: Systems Biology and Medicine</em>, <strong>6</strong>(5), 345-352.
          </li>
          <li ref="ref_1">
            Lu J, Cao X, Zhong S (2018) A likelihood approach to testing hypotheses on the co-evolution of epigenome
            and
            genome.
            <em>PLOS Computational Biology</em>, <strong>14</strong>(12): e1006673. <a href="https://doi.org/10.1371/journal.pcbi.1006673"
              target="_blank">https://doi.org/10.1371/journal.pcbi.1006673</a>
          </li>
        </ol>
      </div>

      <div ref="references" class="section">
        <!-- References -->
        <h1>References</h1>
        <ol>
          <li>Lu J, Cao X, Zhong S (2019) EpiAlignment: alignment with both DNA sequence and epigenomic data.
            <em>bioRxiv</em>, doi: <a
              href="https://doi.org/10.1101/562199"
              target="_blank">https://doi.org/10.1101/562199</a>
          </li>
        </ol>
      </div>
      <div id="footer" class="flexRow flexWrap flexCenter noFlex justifyEnd">
        <div class="flex flexColumn flexStretch flexNoGap">
          <!-- footer text -->
          <div>Copyright © 2018 The Regents of the University of California. All Rights
            Reserved.</div>
          <div class="flex flexRow flexWrap flexStretch flexNoGap">
            <div>Created by Jia Lu, Xiaoyi Cao and Sheng Zhong, Department of
              Bioengineering.</div>
            <a href='&#109;&#097;&#105;&#108;&#116;&#111;:&#115;&#122;&#104;&#111;&#110;&#103;&#064;&#101;&#110;&#103;&#046;&#117;&#099;&#115;&#100;&#046;&#101;&#100;&#117;'>Contact
              us via email</a>
            <a href="https://github.com/Zhong-Lab-UCSD/EpiAlignment_WebApp" target="_blank">Find us on Github</a>
          </div>
        </div>
        <img id="ucsdLogo" src="assets/images/UCSanDiegoLogo-White.png">
      </div>
    </div>
  </div>
</body>

</html>