<!doctype html>
<html>

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>EpiAlignment Manual</title>
  <link href='https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Material+Icons' rel="stylesheet" type="text/css">
  <link href="styles/mainstyles.css" rel="stylesheet" type="text/css">
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-78231712-3"></script>
  <script src="js/googleAnalytics.js"></script>
  <script defer src="https://cdn.jsdelivr.net/npm/vue@2.5.21/dist/vue.min.js"></script>
  <script defer src="js/manual.js" type="module"></script>
</head>

<body class="manual">
  <div id="manual_app" class="flex flexRow flexNoGap flexNoWrap" v-cloak>
    <div v-show="!showNav" id="callOutNav" @click="toggleNav(true)">
      Table of contents
    </div>
    <transition name="nav-slide-right">
      <div v-show="showNav" id="manualNav" class="noFlex flexColumn flexNoGap flexEnd">
        <div id="navHeader" class="flexRow flexWrap flexNoGap flexCenter">
          <div class="flex">Table of contents</div>
          <div id="collapseNavButton" @click="toggleNav(false)">Collapse</div>
        </div>
        <div id="navContent" class="flex">
          <ul>
            <li>
              <a href="#introduction" @click="clickOnNav">Introduction</a>
            </li>
            <li>
              <a href="#quickStart" @click="clickOnNav">Quick start</a>
              <ul>
                <li>
                  <a href="#oneVsOneExample" @click="clickOnNav">One-vs-one mode example</a>
                </li>
                <li>
                  <a href="#manyVsManyExample" @click="clickOnNav">Many-vs-many mode example</a>
                </li>
              </ul>
            </li>
            <li>
              <a href="#usage" @click="clickOnNav">Usage</a>
              <ul>
                <li>
                  <a href="#usageOverview" @click="clickOnNav">Overview</a>
                </li>
                <li>
                  <a href="#oneVsOneMode" @click="clickOnNav">One-vs-one mode</a>
                </li>
                <li>
                  <a href="#manyVsManyMode" @click="clickOnNav">Many-vs-many mode</a>
                </li>
              </ul>
            </li>
            <li>
              <a href="#input" @click="clickOnNav">Input</a>
              <ul>
                <li>
                  <a href="#genomeAssembly" @click="clickOnNav">Genome assembly</a>
                </li>
                <li>
                  <a href="#peakFileAndDatasets" @click="clickOnNav">Peak files and public datasets</a>
                </li>
                <li>
                  <a href="#queryRegions" @click="clickOnNav">Query regions</a>
                </li>
                <li>
                  <a href="#targetRegions" @click="clickOnNav">Target regions</a>
                  <ul>
                    <li>
                      <a href="#oneVsOneModeTarget" @click="clickOnNav">One-vs-one mode</a>
                    </li>
                    <li>
                      <a href="#manyVsManyTarget" @click="clickOnNav">Many-vs-many mode</a>
                    </li>
                  </ul>
                </li>
                <li>
                  <a href="#parameters" @click="clickOnNav">Parameters</a>
                </li>
              </ul>
            </li>
            <li>
              <a href="#output" @click="clickOnNav">Output</a>
              <ul>
                <li>
                  <a href="#oneVsOneOutput" @click="clickOnNav">One-vs-one mode</a>
                </li>
                <li>
                  <a href="#manyVsManyOutput" @click="clickOnNav">Many-vs-many mode</a>
                </li>
              </ul>
            </li>
            <li>
              <a href="#pythonClientTemplate" @click="clickOnNav">Python client template</a>
            </li>
            <li>
              <a href="#furtherReading" @click="clickOnNav">Further reading</a>
            </li>
            <li>
              <a href="#references" @click="clickOnNav">References</a>
            </li>
          </ul>
        </div>
      </div>
    </transition>
    <div id="manualContent" ref="manualContent" class="flex flexColumn flexNoWrap">
      <div id="header" class="noFlex flexRow flexCenter flexWrap">
        <!-- Introduction -->
        <div id="headerContainer" class="flex flexColumn flexStretch flexNoWrap flexNoGap">
          <h1 class="flex"><a href="index.html">EpiAlignment: a chromosomal similarity search engine</a></h1>
          <div class="headerByline">
            EpiAlignment is a tool that can align genomic regions with both sequence
            and epigenomic information.
          </div>
        </div>
        <div id="headerButtonHolder" class="flex flexRow flexWrap flexCenter justifyEnd">
          <a class="buttons flexRow flexCenter justifyCenter" title="A template python program that can be used to access EpiAlignment"
            target="_blank" href="https://github.com/Zhong-Lab-UCSD/EpiAlignment_WebApp/raw/master/html/assets/PythonClient.zip"><i class="material-icons">save_alt</i>Python
            client template</a>
        </div>
      </div>
      <div ref="introduction" class="manualHeader">
        <h1>EpiAlignment Manual</h1>
      </div>
      <div class="section">
        <!--Introduction-->
        <h1>Introduction</h1>
        <p>EpiAlignment is a web-based tool that can align genomic regions with both sequence and epigenomic information.
           This web server aims to assist users with matching genomic elements
           with functional and structural similarities across two species.</p>
        <p>The web server has two alignment modes: one-vs-one (default) and many-vs-many. The default one-vs-one mode is 
           suitable if you have <strong>medium-sized query regions, 
           usually marked by epigenomic modifications, (e.g. ChIP-Seq peaks, hundreds to thousands of base pairs long)</strong>, 
           and much longer target regions. 
           Each query region will be aligned to a sub-region of the target region with the highest similarity.
           The many-vs-many mode attempts to align 
           every input query region to all target regions to find the best hit. 
           It would be useful if you have query and target regions of similar lengths (medium-sized), and would like to find one best match 
           for the query region among a set of target regions. The two modes are illustrated in figures below.</p>
        <div class="flexRow">
          <div class="manualImg small">
            <img alt="Expandable image for one-vs-one mode" src="assets/images/DefaultStart.png">
          </div>
          <div class="manualImg small">
            <img alt="Expandable image for many-vs-many mode" src="assets/images/ManyStart.png">
          </div>
        </div>
      </div>
      <div ref="quickStart" class="section">
        <!--Quick Start-->
        <h1>Quick start</h1>

        <h3 ref="oneVsOneExample">An example for using the one-vs-one mode</h3>
        <!--Example 1-->
        <p>In this example, we will search a H3K4me3-marked mouse genomic region (~2kb) against the neighborhood of its
          homologous region in human. This example will take about 2 minute to run on EpiAlignment.</p>
        <h4>Import sample data with one click:</h4>
        <div class="exampleSteps">
          <ul>
            <li>Make sure that the "One-vs-one" mode is selected in STEP 1.</li>
            <li>Click the "Use sample data" button in STEP 2.</li>
            <li>Click "Submit" in STEP 4</li>
          </ul>
        </div>

        <h4>OR, enter data manually:</h4>
        <div class="exampleSteps">
          <ul class="sampleSteps">
            <li>Make sure that the "One-vs-one" mode is selected in STEP 1.</li>
            <li>In STEP 2, "Query region" box:</li>
            <ul class="sampleStepsSub">
              <li>Set "Genome assembly" to mm10.</li>
              <li>Press "Click here to select a paired public dataset", find "Public Datasets" and select "Round
                spermatids
                ChIP-Seq (H3K4me3)" by clicking on the blue cell.
                Press the confirm button. </li>
              <li>Paste the following mouse genomic region into the "Input regions" textarea:</li>
              <div>
                <pre>chr12&#9;8207583&#9;8209349&#9;Region_example&#9;0&#9;+</pre>
              </div>
            </ul>
            <li>In the "Target region" box, set "Genome assembly" to hg38.</li>
            <li>Make sure that "Use homologous regions in this species" is selected, and the upstream and downstream 
              distances for extension are both 20,000bp.
            </li>
            <li>Click "Submit" in STEP 4.</li>
          </ul>
        </div>

        <h4>Output</h4>
        <p>The result page shows
          an interactive table where each row can be expanded to view detailed results and result evaluations. For visualization, 
          you can click the icon (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">GIVE</span>
          </span>)
          to view query and target regions together 
          in the Genomic Interaction Visualization Engine (GIVE), or view each region individially in the UCSC genome 
          browser by clicking the icon (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">UCSC</span>
          </span>).
          Results can be downloaded and further filtered by clicking "Download results".
          </p>
        <div class="manualImg large">
          <img alt="Expandable data table for one-vs-one mode" src="assets/images/Enhancer_result_comment.png">
        </div>

        <h3 ref="manyVsManyExample">An example for using the many-vs-many mode</h3>
        <!--Example 2-->
        <p>In this example, we will search the promoter of 
          a human protocadherin beta gene, PCDHB5,
          against a mouse gene cluster to find mouse promoters with the
          highest structural similarities
          within it. The mouse gene cluster contains promoters of six mouse protocadherin beta genes: Pcdhb4, Pcdhb6,
          Pcdhb8, Pcdhb10, Pcdhb11 and Pcdhb12, all of which are annotated orthologous genes of PCDHB5.</p>
        <h4>Import sample data with one click:</h4>
        <div class="exampleSteps">
          <ul>
            <li>Select the "Many-vs-many" alignment mode in STEP 1.</li>
            <li>Click the "Use sample data" button in STEP 2.</li>
            <li>Click "Submit" in STEP 4</li>
          </ul>
        </div>
        <h4>OR, enter data manually:</h4>
        <div class="exampleSteps">
          <ul class="sampleSteps">
            <li>Select the "Many-vs-many" alignment mode in STEP 1.</li>
            <li>In STEP 2, "Query region" box:</li>
            <ul class="sampleStepsSub">
              <li>Press "Click here to select a paired public dataset", find "Public Datasets" and select "Round
                spermatids
                ChIP-Seq (H3K4me3)" by clicking on the blue cell.
                Press the confirm button. </li>
              <li>Paste or type the following gene name into the "Input regions"
                textarea, <em>one name per line</em>:</li>
              <div>
                PCDHB5
              </div>
            </ul>
            <li>In the "Target region" box, make sure that "Search a gene cluster"
              is selected, and type "PCDHB5" in the search bar. In the pop-up panel,
              click the box with arrow
              (<i class="material-icons">open_in_browser</i>) to the right of
              the first listed gene cluster <strong>"Cluster_1491"</strong> to select it.</li>
            <li>Click "Submit" in STEP 4.</li>
          </ul>
        </div>

        <h4>Output</h4>
        <p>Two heatmaps, a bar chart and an interactive table will be returned. 
          The heatmaps illustrate alignment scores of every promoter pair 
          and the best hits found for the query promoter. The three digits behind each gene name is the id
          of the promoter region. The bar chart presents expression values of related genes in the same tissue type.</p>
        <div class="manualImg small">
          <img alt="Heatmap for many-vs-many mode" src="assets/images/Heatmap.png">
        </div>
        <p>Similar to that in the default mode, you may expand each row of the table to view more details,
          or click the <i class="material-icons">open_in_new</i> 
          icons to visualize the regions
          in UCSC genome browser or GIVE. The table can be
          downloaded by clicking "Download results".</p>
        <div class="manualImg large">
          <img alt="Data table for many-vs-many mode" src="assets/images/PromoterTable_comments.png">
        </div>
        <p>EpiAlignment also provides many other ways of defining query and target regions. Please check out the Usage
          section for more information.</p>
      </div>
      <div ref="usage" class="section">
        <!--Usage-->
        <h1 ref="usageOverview">Usage</h1>
        <h2>Overview</h2>
        <p>EpiAlignment provides two major alignment modes: the one-vs-one mode and the many-vs-many mode. 
          The default one-vs-one mode attempts to find a best-matching part for a query region in its corresponding target region. 
          It is useful in cases where target regions are much longer than query regions, 
          and best matches of the query regions cannot be easily determined in the target species with sequence homologies only. 
          The many-vs-many mode, on the other hand, was designed to identify the best match for a query region among a set of target regions. 
          It is more useful when the query region corresponds to 
          multiple candidates with potentially similar functions, such as promoters of paralogous genes.</p>
        <div>
          <table>
            <caption>Alignment modes and supported target region types</caption>
            <thead>
              <tr>
                <th>Alignment mode</th>
                <th>Supported target region types</th>
                <th>Description</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <th rowspan="2">One-vs-one mode</th>
                <td>Homologous regions</td>
                <td>For each query region, search for the best-matching sub-region in the neighborhood of its homologous region.</td>
              </tr>
              <tr>
                <td>Genomic coordinates</td>
                <td>For each query region, search for the best-matching sub-region within a designated target region.</td>
              </tr>
              <tr>
                <th rowspan="2">Many-vs-many mode</th>
                <td>A gene cluster</td>
                <td>Align each query region against all isoform promoters in a group of 
                evolutionarily-related genes to identify the best-matching promoter.</td>
              </tr>
              <tr>
                <td>Genomic coordinates or gene names</td>
                <td>Align each query region against all target regions to identify the best match.</td>
              </tr>
            </tbody>
          </table>
        </div>

        <h2 ref="input">Input</h2>
        <!--Input-->
        <p>EpiAlignment requires two types of inputs from users: (1) a pair of peak files from ChIP-Seq experiments, 
          and (2) query and target regions to be aligned, both from two species. 
          The input file formats are different depending on the alignment mode selected.</p>
        <h3 ref="genomeAssembly">Genome assembly</h3>
        <p>Genome assemblies of the query and target regions should first be specified using the dropdown shown below:</p>
        <div class="manualImg small">
          <img alt="Dropdown for genome assemblies" src="assets/images/genomeAssembly.png">
        </div>
        <p>EpiAlignment currently supports two genome assemblies: hg38 (Homo sapiens) and mm10 (Mus musculus). Please
          note
          that the genome assemblies have to be different if you choose to define
          target
          regions using homologous regions in the default mode. For other sub-modes, the two genome assemblies can be
          identical whereas not recommended.</p>

        <h3 ref="peakFileAndDatasets">Peak files and datasets for pairwise ChIP-Seq experiments</h3>
        <p>EpiAlignment provides users with comparable epigenomic datasets of matching tissue/cell types, curated from the RoadMap Epigenomics 
           and ENCODE consortia and public comparative epigenomic studies.To select peak files, 
           in the Query regions box, click "Click here to select a paired public dataset."</p>
        <div class="manualImg">
          <img alt="Public dataset selection button" src="assets/images/presetDataButton.png">
        </div>
        <p>In the pop-up window, click the cell representing the tissue type and histone mark that you want to select.
          There are two numbers on each cell in the format of "H:x M:y". The two numbers represents number of peak
          files
          available in human and mouse, respectively. Once you click the cell, available experiment IDs will be listed
          on
          the
          rightside. You may click the icon beside the experiment ID to see experiment details on their original
          ENCODE/GEO
          database page.</p>
        <p>Click on the experiment ID you'd like to select, then press confirm. The selected experiment
          will display in the Query regions and Target regions boxes afterwards.</p>
        <p>You may also click anywhere outside the pop-up window at any time to close it without making any changes to
          your
          peak file selection.
        </p>
        <div class="manualImg">
          <img alt="Public dataset selection window" src="assets/images/presetWindow.png">
        </div>
        <p>The user can also upload their own peak files. To do this, choose and upload your peak file 
          (<a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1" target="_blank">BED6</a> format)
          for
          the first species in the Query regions box, and the peak file for the second species in the Target regions
          box.
          <em>Uploading your own peak file will clear any previously selected public data in all references.</em></p>
        <!--<div class="manualImg small">
          <img alt="Upload your own peak file" src="assets/images/UploadPeak_comment.png">
        </div>-->

        <h3 ref="queryRegions">
          Query regions
        </h3>
        <p>In both modes, query regions can be provided as genomic coordinates in BED6 format.
          Coordinates can be either pasted into the textarea (one region per line) or uploaded as a file. Typically, the 
          query regions are hundreds to thousands bp-long genomic elements marked by epigenomic modifications, such as ChIP-Seq 
          peaks. If your query regions are longer than 10kb, a warning will show up on the result page.</p>
        <p>Gene symbols or Ensembl IDs are also accepted if gene promoters are used as input regions. 
            Please note that 
            gene symbols are case-sensitive. For human gene symbols, all letters should be uppercase,
            whereas for mouse gene symbols, only the first letter should be uppercase.
          In this case, users need to specify which range around gene transcription start sites (TSSs) will be defined as promoter regions.
          By default, the (-1,000, +500) flanking regions around TSSs are used.</p>
        <div class="manualImg">
          <img alt="Select TSS flanking ranges for promoters" src="assets/images/TSSrange_comments.png">
        </div>

        <p>EpiAlignment will merge different isoform promoters of the same gene 
          if their overlapping region is longer than 80% of the promoter length. For example, if
          promoter regions are 1,500bp long, two promoters will be merged if
          their overlapping region is longer than <span class="equation">1500 &#215; 0.8 = 1200</span>. The last three
          digits of the promoter's genomic coordinate will be used as promoter id.</p>

        <h3 ref="targetRegions">
          Target regions
        </h3>
        <h4 ref="oneVsOneModeTarget">One-vs-one mode</h4>

        <h5>Use homologous regions in this species</h5>
        <p>Query regions will be aligned against their homologous neighborhoods in this sub-mode.
          Users only need to specify the number of base pairs 
          by which the query regions will be increased (default: 20,000bp on each side). 
          The web server will extend each query region and 
          remap it to the target species using <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver" target="_blank">liftOver</a>
          to define the corresponding target region. </p>
        <p>Note: the region pair will not appear in the final result if liftOver fails to remap the query.</p>
        <div class="manualImg small">
          <img alt="Use homologous regions for enhancer targets" src="assets/images/OrthoMode_comments.png">
        </div>

        <h5>Define target regions with a BED file</h5>
        <p>In this sub-mode, target should be provided as genomic coordinates in BED6 format. The list should have the
          same row count and order as the query regions, as EpiAlignment will directly pair them based on their
          line number, and align each pair separately. </p>

        <h4 ref="manyVsManyTarget">Many-vs-many mode</h4>

        <h5>Search a gene cluster</h5>
        <p>In this sub-mode, you may use a set of promoters of evolutionarily-related genes (namely a gene cluster) 
          to define target regions. 
          The gene clusters are preset in EpiAlignment, identified by grouping paralogous genes in each
          species and linking orthologous genes across species. Different isoform promoters of the same genes
          are
          all included. Please note that gene without matching
          paralogues
          in the species are not included in any cluster. 
          An example of a pre-identified gene cluster is shown below, in which <em>Aph1a</em>, <em>Aph1b</em>
          and <em>Aph1c</em> are paralogues in mouse, and <em>APH1A</em>, <em>APH1B</em> are paralogues in human.</p>
        <div>
          <table class="verticalBorder">
            <caption>An example of a pre-identified gene cluster</caption>
            <thead>
              <tr>
                <th></th>
                <th><em>Aph1a</em></th>
                <th><em>Aph1b</em></th>
                <th><em>Aph1c</em></th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <th><em>APH1A</em></th>
                <td>Orthologous</td>
                <td></td>
                <td></td>
              </tr>
              <tr>
                <th><em>APH1B</em></th>
                <td></td>
                <td>Orthologous</td>
                <td>Orthologous</td>
              </tr>
            </tbody>
          </table>
        </div>
        <p>By selecting this cluster, you can either align any human genomic region(s) to the promoters of mouse genes
          <em>Aph1a</em>,
          <em>Aph1b</em>, <em>Aph1c</em>, or any mouse genomic region(s) to the promoters of human genes <em>APH1A</em>,
          <em>APH1B</em>.</p>

        <p>You can type an Ensembl ID or a gene symbol in the searching bar to search a gene cluster. 
          A result panel will pop up with
          candidate clusters listed in it. 
          You may click on the cluster name to further examine genes and their descriptions. 
          Click the box with arrow (<i class="material-icons">open_in_browser</i>) on the rightside of the
          cluster name to select the cluster you would like to
          use. You may click anywhere outside the pop-up panel at anytime to close it.</p>
        <div class="manualImg">
          <img alt="Gene cluster search and selection panel" src="assets/images/GeneClusters_comment.png">
        </div>

        <h5>Define target regions with a BED file / a gene list.</h5>
        <p>In this case, the
          format of the input target regions is exactly the same as query regions. The query and target region files do
          not need to have identical row counts, as EpiAlignment will perform an all-against-all alignment.</p>

        <h3 ref="parameters">Parameters</h3>
        <p>Please refer to <a href="manualParam.html">Choosing parameters in EpiAlignment</a> for details.</p>

        <h2 ref="output">Output</h2>
        <p>In both modes, EpiAlignment will return two types of alignment results, yielded with (<em>w</em> = 0) 
          and without (<em>w</em> = 0) epigenomic information. 
          For each query region, best matches identified in the two alignments 
          are named “EpiAlignment hit” and “Sequence-only hit” respectively.
          All scores are calibrated to 1,000 bp (i.e., alignment score per kilobase), 
          and are thus comparable across alignment results of different region pairs.</p> 
          
        <p>Alignment results are presented in an interactive table on the result page. 
          In both modes, you may click the icon  at the front of each row to launch a browser powered by GIVE 
          (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">GIVE</span>
          </span>)
          to have a side-by-side view of both the query and the target regions. 
          If you have chosen a preset dataset, the dataset will be included in the browser automatically. 
          You may also visualize each region individually in UCSC genome browser by clicking the icon 
          beside each coordinate (<span class="btnWithAnno"><i class="material-icons">open_in_new</i>
            <span class="annotation">UCSC</span>
            </span>). 
          You may select any ENCODE/public epigenomic data tracks in "Custom Tracks", 
          change the display mode from "hide" to "dense" to view them at the same location.</p>

        <p>The table can be sorted by each column by clicking the column name. 
          Users may also click on each row to open an expandable panel and view more details.</p>

        <h3 ref="oneVsOneOutput">One-vs-one mode</h3>
        <p>In this mode, each expandable panel contains a chart presenting alignment scores along the target regions (left) 
          and a result evaluation panel with a number of metrics(right). All metrics are contained in the downloaded table. 
          You may download the table to rank the results by any measurements and select region pairs for future studies 
          (see "Filtering the results" below).</p>
        <ul>
          <li><p><strong>Overview</strong>: the highest alignment scores achieved with and without epigenomic information. 
            A badge saying "The EpiAlignment hit is different from the Sequence-only hit" 
            will be displayed if (1) the coordinates of the two hits don’t overlap, and 
            (2) the EpiAlignment score of the EpiAlignment hit 
            is 10 larger than that of the Sequence-only hit. </p>
            <p>We suggest that the user only consider hits with the badge 
            when looking for cases in which EpiAlignment and Sequence-only alignment results don't agree. </p></li>     
          <li>
            <p><strong>Sequence evaluation</strong>: an evaluation of sequence similarities between the query region and the two hits. 
            For every hit, a signal-to-noise ratio (SNR) is provided comparing its sequence alignment score (<em>S</em> score)
            to the scores of arbitrary locations within the target region. 
            An SNR no larger than 1 indicates that the hit’s sequence similarity to the query is indistinguishable from random sequences nearby.</p>
            <p>Two reference distributions of <em>S</em> scores are also provided for comparison purposes. The distribution of conserved mouse-human region pairs 
            was generated using mouse regions marked by H3K4me3, H3K27ac or H3K4me1 in any tissue type included in our ChIP-Seq database and their human 
            homologues, identified using liftOver (minMatch = 0.5). These mouse regions were also aligned to randomly-selected human genomic regions (~6kb long) to generate the background distribution. 
            Approximately 15,000 mouse-human region pairs were randomly-selected and aligned to generate each distribution. 
            In general, sequence similarity scarcely contributes to the alignment when SNR close to 1 or <em>S</em> score close to the background.</p></li>
          <li><p></p><strong>Epigenome evaluation</strong>: an evaluation of epigenomic similarities between the query region and the two hits.
            A hit's epigenomic similarity score <em>E</em> is defined as the difference between its alignment scores yielded with and without epigenomic information.
            A theoretical range of <em>E</em> 
            is provided based on the query epigenomic profile, 
            where the maximum and the minimum can be achieved when all epigenomic states are rewarded or penalized, respectively. 
            A larger <em>E</em> value suggests higher epigenomic similarity between two regions.</p>
            <p>Please note that a negative epigenomic similarity score does not mean that one region 
            is marked by epigenomic modification while the other is not. The query region and the hit may both contain epigenomic peaks, 
            whereas the peak regions barely overlap.</p></li>  
          <li><p><strong>Expression values of surrounding genes</strong>: expression levels of genes within 100kb flanking regions 
            of the query and target regions. This part will only be displayed if RNA-Seq data of the same tissue type is available.</p></li>   
        </ul>
        
        <p>Note: if you set epigenomic weight to zero during input, the EpiAlignment score is equivalent to
          sequence-only alignment score, and no "Sequence-only" result will be reported.</p>
        <div class="manualImg">
          <img alt="Expandable data table for one-vs-one mode" src="assets/images/oneManualResult.png">
        </div>

        <h3 ref="manyVsManyOutput">Many-vs-many mode</h3>
        <p>In many-vs-many mode, each expandable panels contains two reference distributions of sequence alignment scores (<em>S</em> score)
          for the evaluation of sequence similarity between two promoters. The background distribution was generated by aligning 
          10,000 pairs of randomly-paired human and mouse regions. 
          The <em>S</em> score distribution of orthologous promoter pairs was generated using all promoter pairs whose corresponding 
          human and mouse genes are unique orthologues to each other. Only genes included in the database for 
          evolutionarily-related genes were used (4,615 pairs in total).</p>
        <div class="manualImg">
          <img alt="Heatmap for many-vs-many mode" src="assets/images/Promoter_table_expand.png">
        </div>
        <p>Alignment scores between query and target regions are illustrated in the two heatmaps on the result page, 
          where rows correspond to query regions, and columns correspond to target regions.
          In each row, the cell with the highest score of the row is marked with a
          capital "H".</p>
        <p>For each query region (row), EpiAlignment detects if its best hit switches from one target region to another
          when aligning with and without epigenomic information. The best hit is considered as "changed" if it meets
          the following criteria:</p>



        <div class="exampleSteps">
          <div class="flexRow flexWrap flexStretch">
            <table>
              <caption>Alignment scores with epigenomic information</caption>
              <thead>
                <tr>
                  <th></th>
                  <th>Target region A</th>
                  <th>Target region B</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th>Query region</th>
                  <td><span class="equation"><em>T<sub>A</sub></em></span></td>
                  <td><span class="equation"><em>T<sub>B</sub></em></span></td>
                </tr>
              </tbody>
            </table>
            <table>
              <caption>Alignment scores without epigenomic information</caption>
              <thead>
                <tr>
                  <th></th>
                  <th>Target region A</th>
                  <th>Target region B</th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <th>Query region</th>
                  <td><span class="equation"><em>S<sub>A</sub></em></span></td>
                  <td><span class="equation"><em>S<sub>B</sub></em></span></td>
                </tr>
              </tbody>
            </table>
          </div>
          <p>Assume that in alignment <em>without</em> epigenomic information,
            <span class="equation"><em>S<sub>A</sub></em> &#62; <em>S<sub>B</sub></em></span>,
            whereas in alignment <em>with</em> epigenomic information,
            <span class="equation"><em>T<sub>A</sub></em> &#60; <em>T<sub>B</sub></em></span>. For
            the query region, the best hit of the query
            region will be considered as switched from
            Target A to B when using EpiAlignment if <span class="equation"><em>T<sub>B</sub></em> &#62; <em>S<sub>B</sub></em></span>
            and <span class="equation"><em>T<sub>A</sub></em> &#60; <em>S<sub>A</sub></em></span>.
            In other words, we required that for the hit found by EpiAlignment with epigenomic information,
            the epigenomic information has to reinforce the chromosomal similarity, where as for the hit found by
            sequence-only alignment, the epigenomic information has to undermine the chromosomal similarity.</p>
        </div>

        <p>For query regions with switched target, the best hit found by EpiAlignment with epigenomic information
          is highlighted by a green box, and the best hit found by alignment without epigenomic information is
          highlighted by a gray box. </p>
        
        <p>If input regions are defined with gene symbols or IDs, 
          the gene expression levels will be presented in a bar chart next to the heatmaps when the data is available.
          You may compare the gene expression data with EpiAlignment results side-by-side to further investigate 
          potential consequences of epigenomic similarities.</p>

        <div class="manualImg large">
          <img alt="Heatmap for many-vs-many mode" src="assets/images/Heatmap.png">
        </div>
      </div>

      <div ref="pythonClientTemplate" class="section">
        <h1>Python client template</h1>
        <p>We provided EpiAlignment_PyClient.py, a Python template program for users to access EpiAlignment from their
          own server, submit multiple jobs and parse the output. The user may also modify the code and incorporate it
          into their own programs. The Python client template can be downloaded from the homepage of EpiAlignment.
          Sample input files are also available in the same zip file. Please follow <a
            href="https://github.com/Zhong-Lab-UCSD/EpiAlignment_WebApp/blob/master/html/assets/PythonClient/README.md"
            target="_blank">the README.md file</a> (also included in the zip file) to prepare your sample sheet and
          submit jobs to EpiAlignment.</p>
        <p>You may run the Python client via the following command:</p>
        <code class="codeParagraph">python EpiAlignment_PyClient.py sampleSheet.txt</code>
        <p>where sampleSheet.txt is a tab-delimited text file with details of the jobs you would like to submit to
          EpiAlignment.</p>
        <h3>Running the client template with <code>pipenv</code></h3>
        <p>If <a href="https://pipenv.readthedocs.io/" target="_blank"><code>pipenv</code></a> is available, you may
          initialize the running environment with the following command:
        </p>
        <code class="codeParagraph">pipenv install</code>
        <p>Then you may run the client within the initialized environment by:</p>
        <code class="codeParagraph">pipenv run python EpiAlignment_PyClient.py sampleSheet.txt</code>
      </div>
      <!--<div class="section">
        <h1>Computational methods</h1>
        <p>EpiAlignment uses a dynamic programming algorithm for the comparison of sequences and epigenomic modifications. The input data is a pair of genomic regions: the query region <i>A</i> and the target region <i>B</i>, each of which consists of the DNA sequence and the epigenomic modifications on it. We assume that the state for each epigenomic modification on each genomic location is binary as shown in the figure below. On each base, a binary indicator (1,0) is used to represent the epigenomic state of an epigenomic mark, with the epigenomic state equal to 1 within regions marked by the epigenomic mark, and 0 outside such regions. </p>
    
        <p>EpiAlignment starts from scoring a pair of nucleoside bases from two sequences, rewarding for matches and penalizing for mismatches, gap openings or gap extensions. On top of each pair of bases, EpiAlignment adds a score from the epigenomic states to the alignment score, rewarding for matches and penalizing for mismatches of each epigenomic mark. Through dynamic programming, EpiAlignment computes alignment scores of any pairs of sub-sequences, and reports the path with the highest alignment score. </p>
        
      </div> -->

      <div ref="furtherReading" class="section">
        <!-- Further reading -->
        <h1>Further Reading</h1>
        <ol>
          <li>
            Xiao, S., Xie, D., Cao, X., Yu, P., Xing, X., Chen, C. C., . . . Zhong, S. (2012). Comparative epigenomic annotation of regulatory DNA. <em>Cell</em>, <strong>149</strong>(6), 1381-1392.
          </li>
          <li>
            Xiao, S., Cao, X., &amp; Zhong, S. (2014). Comparative epigenomics: defining and utilizing epigenomic variations across species, time‐course, and individuals. <em>Wiley Interdisciplinary Reviews: Systems Biology and Medicine</em>, <strong>6</strong>(5), 345-352.
          </li>
          <li ref="ref_1">
            Lu J, Cao X, Zhong S (2018) A likelihood approach to testing hypotheses on the co-evolution of epigenome
            and
            genome.
            <em>PLOS Computational Biology</em>, <strong>14</strong>(12): e1006673. <a href="https://doi.org/10.1371/journal.pcbi.1006673"
              target="_blank">https://doi.org/10.1371/journal.pcbi.1006673</a>
          </li>
        </ol>
      </div>

      <div ref="references" class="section">
        <!-- References -->
        <h1>References</h1>
        <ol>
          <li>Lu J, Cao X, Zhong S (2019) EpiAlignment: alignment with both DNA sequence and epigenomic data.
            <em> Nucleic Acids Res</em>, 47(W1), W11-W19. 
          </li>
        </ol>
      </div>
      <div id="footer" class="flexRow flexWrap flexCenter noFlex justifyEnd">
        <div class="flex flexColumn flexStretch flexNoGap">
          <!-- footer text -->
          <div>Copyright © 2018 The Regents of the University of California. All Rights
            Reserved.</div>
          <div class="flex flexRow flexWrap flexStretch flexNoGap">
            <div>Created by Jia Lu, Xiaoyi Cao and Sheng Zhong, Department of
              Bioengineering.</div>
            <a href='&#109;&#097;&#105;&#108;&#116;&#111;:&#115;&#122;&#104;&#111;&#110;&#103;&#064;&#101;&#110;&#103;&#046;&#117;&#099;&#115;&#100;&#046;&#101;&#100;&#117;'>Contact
              us via email</a>
            <a href="https://github.com/Zhong-Lab-UCSD/EpiAlignment_WebApp" target="_blank">Find us on Github</a>
          </div>
        </div>
        <img id="ucsdLogo" src="assets/images/UCSanDiegoLogo-White.png">
      </div>
    </div>
  </div>
</body>

</html>